src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,770 @@
+			case '-TT':
+			case '-U':
+				defined( 'DEBUG' ) && printf( "Encountering \e[31m{$tag_name}\e[m closer at \e[35m{$this->current_token->bookmark_name}\e[m\n" );
+				$this->run_adoption_agency_algorithm();
+				return true;
+
+			/*
+			 * > A start tag whose tag name is one of: "area", "br", "embed", "img", "keygen", "wbr"
+			 */
+			case '+IMG':
+				$this->reconstruct_active_formatting_elements();
+				$this->insert_html_element( $this->current_token );
+				return true;
+
+			default:
+				throw new WP_HTML_Unsupported_Exception( "Cannot process {$tag_name} element." );
+		}
+	}
+
+	/*
+	 * Internal helpers
+	 */
+
+	/**
+	 * Creates a new bookmark for the currently-matched tag.
+	 *
+	 * @return false|string Name of created bookmark, or false if unable to create.
+	 */
+	private function bookmark_tag() {
+		if ( ! $this->get_tag() ) {
+			return false;
+		}
+
+		if ( ! parent::set_bookmark( ++$this->bookmark_counter ) ) {
+			defined( 'DEBUG' ) && printf( "\e[43mCouldn't set bookmark \e[44m\e[37m{$this->bookmark_counter}\e[m!\n" );
+			return false;
+		}
+
+		return "{$this->bookmark_counter}";
+	}
+
+	/*
+	 * HTML semantic overrides for Tag Processor
+	 */
+
+	/**
+	 * Returns the uppercase name of the matched tag.
+	 *
+	 * Example:
+	 *
+	 *     $p = new WP_HTML_Tag_Processor( '<div class="test">Test</div>' );
+	 *     $p->next_tag() === true;
+	 *     $p->get_tag() === 'DIV';
+	 *
+	 *     $p->next_tag() === false;
+	 *     $p->get_tag() === null;
+	 *
+	 * @since {WP_VERSION}
+	 *
+	 * @return string|null Name of currently matched tag in input HTML, or `null` if none found.
+	 */
+	public function get_tag() {
+		$tag_name = parent::get_tag();
+
+		switch ( $tag_name ) {
+			case 'IMAGE':
+				/*
+				 * > A start tag whose tag name is "image"
+				 * > Change the token's tag name to "img" and reprocess it. (Don't ask.)
+				 */
+				return 'IMG';
+
+			default:
+				return $tag_name;
+		}
+	}
+
+	public function release_bookmark( $bookmark_name ) {
+		return parent::release_bookmark( "_{$bookmark_name}" );
+	}
+
+	public function seek( $bookmark_name ) {
+		return parent::seek( "_{$bookmark_name}" );
+	}
+
+	public function set_bookmark( $bookmark_name ) {
+		return parent::set_bookmark( "_{$bookmark_name}" );
+	}
+
+	/*
+	 * HTML Parsing Algorithms
+	 */
+
+	/**
+	 * Closes a P element.
+	 *
+	 * @see https://html.spec.whatwg.org/#close-a-p-element
+	 *
+	 * @return void
+	 */
+	private function close_a_p_element() {
+		$this->generate_implied_end_tags( 'P' );
+		$this->state->stack_of_open_elements->pop_until( 'P' );
+	}
+
+	/**
+	 * Closes elements that have implied end tags.
+	 *
+	 * @see https://html.spec.whatwg.org/#generate-implied-end-tags
+	 *
+	 * @param string|null $except_for_this_element Perform as if this element doesn't exist in the stack of open elements.
+	 * @return void
+	 * @throws WP_HTML_Unsupported_Exception
+	 */
+	private function generate_implied_end_tags( $except_for_this_element = null ) {
```

Comment:
I kind of dislike following the naming from the spec because we don't actually put end tags anywhere, only altering the parser state. Using the same language is a worthy trade-off, though 🤷

Reply:
they are implied 😆 

joking aside you convinced me to follow the spec naming. somewhere down the line is my split-stack approach that actually does generate the implied end tags - this is still a concern I have with the stateful stack of open elements. yes, when building a DOM tree as we go we don't need to have the end tags, but when maintaining the string and needing to jump back and forth they can be very useful